File: backend/apps/users/validators.py
from django.core.exceptions import ValidationError
import re

class StrongPasswordValidator:
    """Custom validator matching the frontend password rules."""

    def validate(self, password, user=None):
        errors = []
        if len(password) < 8:
            errors.append("Password must be at least 8 characters long.")
        if not re.search(r'[A-Z]', password):
            errors.append("Password must contain at least one uppercase letter.")
        if not re.search(r'[a-z]', password):
            errors.append("Password must contain at least one lowercase letter.")
        if not re.search(r'[0-9]', password):
            errors.append("Password must contain at least one number.")
        if not re.search(r'[!@#$%^&*()\,.\?":{}|<>]', password):
            errors.append("Password must contain at least one special character (!@#$%^&*).")
        if errors:
            raise ValidationError(errors)

    def get_help_text(self):
        return (
            "Password must be at least 8 characters long and include "
            "uppercase, lowercase, a number, and a special character."
        )

File: frontend-v2/src/utils/validators.js
export const PASSWORD_RULES = [
  { id: "length", label: "At least 8 characters", test: (p) => p.length >= 8 },
  { id: "uppercase", label: "At least 1 uppercase letter (A-Z)", test: (p) => /[A-Z]/.test(p) },
  { id: "lowercase", label: "At least 1 lowercase letter (a-z)", test: (p) => /[a-z]/.test(p) },
  { id: "digit", label: "At least 1 digit (0-9)", test: (p) => /[0-9]/.test(p) },
  { id: "special", label: "At least 1 special character (!@#$%...)", test: (p) => /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(p) },
];

export const checkPasswordStrength = (password) => {
  return PASSWORD_RULES.every((rule) => rule.test(password || ""));
};
